substratevm/src/com.oracle.svm.core.graal.llvm/src/com/oracle/svm/core/graal/llvm/SubstrateLLVMBackend.java

```diff
@@ -121,6 +125,14 @@ private void emitLLVM(StructuredGraph graph, CompilationResult result) {
             NodeLLVMBuilder nodeBuilder = newNodeLLVMBuilder(graph, generator);
 
             /* LLVM generation */
+            if (LLVMOptions.IncludeLLVMSourceDebugInfo.getValue()) {
+                //TODO: Avoid this lock if possible
+                imageSingletonesLock.lock();
+                if (ImageSingletons.contains(SourceManager.class) == false) {
```

Comment:
No, DebugInfoFeature is the right place to do this as it is relevant to all debug info generation code not just the LLVM debug info generator.

Reply:
Yes, `SourceManager` is definitely not thread safe at the moment.

Reply:
I agree, having all management related to `SourceManager` in `NativeImageDebugInfoFeature` makes sense. The handling of debug information on the LLVM backend has to happen in the compilation phase, so making `SourceManager` can't be avoided, I think. It should be as simple as changing its data structures to thread-safe ones though, so not a big hurdle.

Reply:
Well, this is a cache manager so it is not quite so simple as just updating the data structures. The important thing to get right is to ensure that the check as to whether a source file is cached in the sources directory and any subsequent lookup and copy from some originating jar into the sources directory happens *atomically*. A secondary constraint is that a failed lookup must records the lack of any suitable source atomically so that we don't waste time repeating lookups.

Like all concurrent caches there are different choices available to enforce consistency, ranging from the sledgehammer solution of a synchronization on the whole cache at every lookup, through a synchronization on some proxy that exists per cached file, right down to a racy lookup and copy of sources that only synchronizes at the point of linking the copied file into the source tree.

I am happy to review any proposed fix. However, I am also happy to implement a solution, which may be quicker as I am already familiar with the code. @rishikeshdevsot @loicottet please confirm how you would like to proceed.

Reply:
@adinn I am okay with you implementing a solution since you are familiar with the problems and the code

Reply:
Ok, I'll prepare that in a separate PR and you can rebase this PR on it.